indexing-service/src/test/java/io/druid/indexing/overlord/RemoteTaskRunnerRunPendingTasksConcurrencyTest.java

```diff
@@ -0,0 +1,150 @@
+/*
+ *
+ *  Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ *  or more contributor license agreements. See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership. Metamarkets licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License. You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied. See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ * /
+ *
+ */
+
+package io.druid.indexing.overlord;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import io.druid.indexing.common.TaskStatus;
+import io.druid.indexing.common.TestTasks;
+import io.druid.indexing.common.task.Task;
+import org.joda.time.Period;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ */
+public class RemoteTaskRunnerRunPendingTasksConcurrencyTest
+{
+  private RemoteTaskRunner remoteTaskRunner;
+  private RemoteTaskRunnerTestUtils rtrTestUtils = new RemoteTaskRunnerTestUtils();
+
+  @Before
+  public void setUp() throws Exception
+  {
+    rtrTestUtils.setUp();
+  }
+
+  @After
+  public void tearDown() throws Exception
+  {
+    if (remoteTaskRunner != null) {
+      remoteTaskRunner.stop();
+    }
+    rtrTestUtils.tearDown();
+  }
+
+  // This task reproduces the races described in https://github.com/druid-io/druid/issues/2842
+  @Test(timeout = 90_000)
+  public void testConcurrency() throws Exception
```

Comment:
how much time does this test takes to complete ?
there are thread.sleep in this test, is there any way we can use barriers or latches instead of sleep ?